### ISSUE TITLE
Add option to match relative URLs without computing isIndirectPathRequest()

### ISSUE THREAD

**Issue by Caerbannog** [httptoolkit/mockttp#30]
The following commit prevents us from upgrading: https://github.com/httptoolkit/mockttp/commit/ef2e0a88574afa8fb37dac834dcecc67e8a21426
> This may be a breaking change if you are sending requests to a
Mockttp server directly, using a hostname for the server that
isn't 'localhost' or one of its ip addresses. That traffic will
now be treated as having an absolute URL.

We will be able to get around the issue, but I thought that describing our use case could help you if you wanted to change the design.

Our use case is system testing of an `app` container which sometimes makes requests to an `external_webservice`.
Our tests are run by another container named `tester`, which makes requests to the `app` container and checks the response. We have been using mockttp to mock `external_webservice`. A typical test inside `tester` would be:

```js
describe('The APP container on /route1', () => {
  before(async () => {
    await mockServer.start(process.env.MOCK_PORT);
    await mockServer.get('/external_api').thenReply(200, "something that external_webservice should respond");
  });
  after(async () => {
    await mockServer.stop();
  });
  it('always calls the external_api and returns OK', async () => {
    const response = await request(`${process.env.APP_URL}/route1`);
    expect(response).to.equal('OK');
    const externalRequests = await mockServer.getSeenRequests();
    expect(externalRequests.length).to.equal(1);
  });
}
```

Our containers run with docker-compose, with something like this;
```
app:
  environment:
    EXTERNAL_API_URL: http://tester:7500/external_api
tester:
  environment:
    MOCK_PORT: 7500
```

As you can see, the host that mockttp will receive is `tester` not `localhost`.
After upgrading, this is the error message:
```
No rules were found matching this request.
This request was: GET request to http://tester:7500/external_api with headers: ...
The configured rules are:
Match requests making GETs for /external_api, and then respond with status 200 and body "something that external_webservice should respond".
You can fix this by adding a rule to match this request, for example:
mockServer.get("/external_api").thenReply(200, "your response");
```

Note that the suggestion is incorrect, because this is precisely the route that we specified.

We can get around the issue in multiple ways. But maybe mockttp could be improved in one of the following ways:
- add a `disableProxy` option to `MockttpOptions`
- add a `noHostChecking` option to `MockttpOptions`
- create a `mockServer.proxyUrl`, which would be different from `mockServer.url` (maybe appending a fingerprint like `/.well-kown/mockttp-proxy`)

**Comment by pimterry** [comment]
Interesting! That's very useful, thanks, I wasn't sure whether this case was one that anybody would hit - clearly it is :-).

In the short term, you should be able to match this request by changing your rule to use the absolute form:

```js
await mockServer.get('http://tester:7500/external_api').thenReply(...);
```

Does that work for you?

You could inject `tester:7500` through an env var if you'd prefer, or alternatively make this independent of host entirely with a regex (`get(/https?:\/\/[^\/]+\/external_api/)`).

In the medium term, I agree this makes that case difficult, and it is a case I'd like to support properly. That said, I'm cautious of solving it with new configuration options or by separating the proxied & direct request cases unnecessarily, since they're fundamentally very similar, and it's occasionally even useful to do both at the same time.

There's an argument to be made that the issue here is simply in how path matching works perhaps? This could be changed so that absolute URL matchers check the host, whilst any relative URLs only check the path etc, and match for _all_ hosts.

That would mean the host information is still available for matching when that's useful, but the simpler option still works too, which seems fairly intuitive to me.

Given that, both `get('http://tester:7500/external_api')` and `get('/external_api')` would match. Currently only the former does, and previously only the latter.

What do you think?

**Comment by Caerbannog** [comment]
> await mockServer.get('http://tester:7500/external_api').thenReply(...);
> Does that work for you?

Yes, it works.

> I'm cautious of solving it with new configuration options or by separating the proxied & direct request cases

Right.

> This could be changed so that absolute URL matchers check the host, whilst any relative URLs only check the path etc, and match for all hosts.
> ...
> What do you think?



That's a great idea!

**Comment by pimterry** [comment]
@Caerbannog I've taken a first pass at this in https://github.com/httptoolkit/mockttp/commit/415405441fea5b906a5caeded720ca6a562421ba. Both regex & path matching can now matches against either the path or the absolute URL, i.e. you can match a GET request to `http://localhost:8000/abc` using any of:

* `server.get('/abc')`
* `server.get('localhost:8000/abc')`
* `server.get('http://localhost:8000/abc')`
* `server.get(/\/abc/)`
* `server.get(/http:\/\/localhost:8000\/abc/)`

Note that as part of this, I'm changing `request.url` to be _always_ absolute. This simplifies things quite a bit internally, and makes it more consistent externally, but might affect your tests, if you check for a specific `.url` value on a request returned by `getSeenRequests()`, for example.

Would this work for you?

This is obviously a breaking change, so I'm intending to release it as 0.17.0. I'll do that once I've done a little more testing & wrapped up a couple of other bits along the way, but it shouldn't be too far off.

**Comment by pimterry** [comment]
@Caerbannog, any thoughts on this?

**Comment by Caerbannog** [comment]
From reading your messages it seems great, but I haven't had the time to test the commit in our codebase.

**Comment by Caerbannog** [comment]
Today I tried installing the `master` branch of mockttp to test these changes.
I failed, probably because I have too little experience with typescript to debug trivial issues.
I was trying to reproduce this build: https://cdn.travis-ci.org/httptoolkit/mockttp/jobs/570280930

Do you think that writing a short guide to build mockttp would be usefull for other contributors?

The errors I started with:

> custom-typings/http-proxy-agent.d.ts:2:27 - error TS2307: Cannot find module 'http'.
> src/util/util.ts:17:77 - error TS2580: Cannot find name 'process'. Do you need to install type definitions for node? Try `npm i @types/node`.
> ...
> 17 export const isNode = !isWorker && !isWeb && typeof process === 'object' && process.version;
>                                                                                ~~~~~~~
> Found 282 errors.
> npm ERR! code ELIFECYCLE
> npm ERR! errno 1
> npm ERR! mockttp@0.16.1 build:src: `tsc && cp src/standalone/schema.gql 
> dist/standalone/schema.gql && chmod +x ./dist/standalone/standalone-bin.js`
> npm ERR! Exit status 1

The error I ended with and could not fix was:
> Namespace 'Chai' has no exported member 'ChaiPlugin'.

**Comment by pimterry** [comment]
That's surprising @Caerbannog, not sure what's happened there. What commands did you run exactly?

Travis effectively just runs:

* `npm install`
* `npm test`

In theory running that in a clean checkout should install all dependencies, build the code & test it, with no other setup required.

What does `node --version` and `npm --version` show?

**Comment by Caerbannog** [comment]
I retried it on a mostly clean host:
```
ubuntu@host:~/mockttp$ node --version
v9.4.0
ubuntu@host:~/mockttp$ npm --version
5.6.0
ubuntu@host:~/mockttp$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
ubuntu@host:~/mockttp$ npm test

> mockttp@0.16.1 test /home/ubuntu/mockttp
> npm run build && npm run test:node && npm run test:browser


> mockttp@0.16.1 prebuild /home/ubuntu/mockttp
> rimraf dist/*


> mockttp@0.16.1 build /home/ubuntu/mockttp
> npm run build:src && npm run build:doc


> mockttp@0.16.1 build:src /home/ubuntu/mockttp
> tsc && cp src/standalone/schema.gql dist/standalone/schema.gql && chmod +x ./dist/standalone/standalone-bin.js

node_modules/@types/chai-as-promised/index.d.ts:14:43 - error TS2694: Namespace 'Chai' has no exported member 'ChaiPlugin'.

14     interface ChaiAsPromised extends Chai.ChaiPlugin {
                                             ~~~~~~~~~~


Found 1 error.

npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! mockttp@0.16.1 build:src: `tsc && cp src/standalone/schema.gql dist/standalone/schema.gql && chmod +x ./dist/standalone/standalone-bin.js`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the mockttp@0.16.1 build:src script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/2019-08-13T16_06_48_663Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! mockttp@0.16.1 build: `npm run build:src && npm run build:doc`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the mockttp@0.16.1 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/ubuntu/.npm/_logs/2019-08-13T16_06_48_678Z-debug.log
npm ERR! Test failed.  See above for more details.
```

**Comment by pimterry** [comment]
Got it, good find @Caerbannog!

Unfortunately `@types` packages generally don't follow semver. They try to match version numbers to the corresponding real package, so major type changes (hard to really define anyway) can't result in a major `@types` package bump. The general solution to fixing that is to not have floating `@types` dependencies and instead pin them to a specific version.

Unfortunately at some point some of these dev dependencies become unpinned, and there's a new version of `@types/chai-as-promised` that was released very recently, which breaks the build due to some assumptions they're making about other package versions.

I've now pinned all the type dependencies to the right versions. If you pull the latest commit from master and reinstall you should be all good, and now that they're pinned this shouldn't happen again. Sorry about that! Let me know if that works.

**Comment by Caerbannog** [comment]
Thank you, this worked.
I have tested master with one of my repositories that used to fail, and now tests are passing.
So the issue is probably resolved.

**Comment by pimterry** [comment]
Great! Thanks for testing that.

I'm doing a last bit of extra changes at the moment - I'll close this ticket once the new release is ready, should be just a few days away.

**Comment by pimterry** [comment]
This has now been released as v0.17.0, so I'll close this.

Thanks again for the report, discussion & testing, this was really useful! Let me know if you have any more issues in future.
